gate-web/src/main/groovy/com/netflix/spinnaker/gate/controllers/ApiExtensionController.java

```diff
@@ -0,0 +1,81 @@
+package com.netflix.spinnaker.gate.controllers;
+
+import static org.springframework.web.bind.annotation.RequestMethod.*;
+
+import com.netflix.spinnaker.gate.api.extension.ApiExtension;
+import com.netflix.spinnaker.gate.api.extension.HttpRequest;
+import com.netflix.spinnaker.gate.api.extension.HttpResponse;
+import com.netflix.spinnaker.kork.annotations.Alpha;
+import com.netflix.spinnaker.kork.exceptions.SpinnakerException;
+import com.netflix.spinnaker.kork.web.exceptions.NotFoundException;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import javax.servlet.http.HttpServletRequest;
+import okhttp3.internal.http.HttpMethod;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RestController;
+
+/**
+ * A top-level {@code RestController} that exposes all api extensions under a common
+ * `/extensions/{{ApiExtension.id}}` path.
+ */
+@Alpha
+@RestController
+@RequestMapping("/extensions")
+public class ApiExtensionController {
+  private List<ApiExtension> apiExtensions;
+
+  @Autowired
+  public ApiExtensionController(List<ApiExtension> apiExtensions) {
+    this.apiExtensions = apiExtensions;
+  }
+
+  @RequestMapping(value = {"/{extension}/**"})
+  public ResponseEntity<Object> any(
+      @PathVariable(value = "extension") String extension,
+      @RequestParam Map<String, String> requestParams,
+      HttpServletRequest httpServletRequest) {
+    HttpRequest httpRequest =
+        HttpRequest.of(
+            httpServletRequest.getMethod(),
+            httpServletRequest.getRequestURI().replace("/extensions/" + extension, ""),
+            requestParams);
+
+    if (HttpMethod.permitsRequestBody(httpRequest.getMethod())) {
+      try {
+        httpRequest.setBody(
+            httpServletRequest
+                .getReader()
+                .lines()
+                .collect(Collectors.joining(System.lineSeparator())));
+      } catch (IOException e) {
+        throw new SpinnakerException("Unable to read request body", e);
+      }
+    }
+
+    Optional<ApiExtension> apiExtension =
+        apiExtensions.stream()
```

Comment:
I think `apiExtensions` could be null.  May want to make this an Optional stream (either something like `Optional.ofNullable` or making the `apiExtensions` field optional).

Reply:
I did verify that the constructor receives an empty list (rather than a null list) when there are no api extensions present.

Reply:
Ah, cool :+1:

Reply:
If asked to provide a `Collection<T>`, Spring will never provide `null` if there are no beans of `T`. 🙏